generate_specs.py

```diff
@@ -1,89 +1,144 @@
 """Script to generate specifications for C functions using an LLM and verify them with CBMC."""
 
+from __future__ import annotations
+
+import re
 import subprocess
 import sys
 from pathlib import Path
 
-from analysis import CodeGraph
+from analysis import CallGraph
 from models import LLMGen, get_llm_generation_with_model
-from util import LLVMAnalysis, extract_specifications
+from util import ParsecResult, PromptBuilder, extract_specification
 
 MODEL = "gpt-4o"
 HEADERS_TO_INCLUDE_IN_OUTPUT = ["stdlib.h", "limits.h"]
+DEFAULT_NUM_VERIFICATION_ATTEMPTS = 10
+
+
+def main() -> None:
+    """Entry point to generate_specs.py."""
+    input_file_path = Path(sys.argv[1])
+    output_file_path = _prepare_output_location(input_file_path)
+
+    call_graph = CallGraph(output_file_path)
+    recursive_funcs = call_graph.get_names_of_recursive_functions()
+
+    # TODO: Process recursive loops rather than removing them. A topological ordering is not
+    # computed in cases where recursion is present, and we fall back to a DFS post-order traversal.
+    # Note: No recursive functions are actually removed from the graph; only self-edges.
+    call_graph_without_self_edges = call_graph.remove_self_edges()
```

Comment:
This line and the assignment to `recursive_funcs` seem to do closely related tasks.  Should they be combined?